### DISCUSSION TITLE
the roles in chat_history seem confused?

### DISCUSSION THREAD

**Discussion by huangch** [microsoft/autogen#1839] [Q&A] [answered]
Hi, I'm looking into the notebook https://github.com/microsoft/autogen/blob/main/notebook/agentchat_function_call_currency_calculator.ipynb

In the last cell:
print("Chat history:", res.chat_history)
Chat history: [{'content': 'How much is 123.45 US Dollars in Euros?', 'role': 'assistant'}, {'tool_calls': [{'id': 'call_qv2SwJHpKrG73btxNzUnYBoR', 'function': {'arguments': '{"base":{"currency":"USD","amount":123.45},"quote_currency":"EUR"}', 'name': 'currency_calculator'}, 'type': 'function'}], 'content': None, 'role': 'assistant'}, {'content': '{"currency":"EUR","amount":112.22727272727272}', 'tool_responses': [{'tool_call_id': 'call_qv2SwJHpKrG73btxNzUnYBoR', 'role': 'tool', 'content': '{"currency":"EUR","amount":112.22727272727272}'}], 'role': 'tool'}, {'content': '123.45 US Dollars is approximately 112.23 Euros.', 'role': 'user'}, {'content': '', 'role': 'assistant'}, {'content': 'TERMINATE', 'role': 'user'}]

I'm wondering, the the first message, "How much is 123.45 US Dollars in Euros?", is supposedly asked by the user. But in the chat_history, it showed that the role is the assistant?

In the first response, the program shows this message, "How much is 123.45 USD in EUR?" was "user_proxy (to chatbot):".

Did I have misunderstood the meaning of the 'role' in the chat_history?

**Comment by rickyloynd-microsoft** [comment, marked as answer]
This is a common source of confusion. When two agents are conversing, each maintains its own copy of the message history, and their roles are reversed in those two copies. The reason for this is because openai defined a fixed set of message roles like user and assistant, and autogen follows those roles in its message lists. In a group chat it may be that none of the agents represent the user, and yet every message list sent to the LLM must be annotated with user and assistant roles. The simple rule to remember is that every agent carries the assistant role in its own copy of the message list.

**Reply by huangch** [reply]
Thank you. But how do you keep track of the roles after the conversation finished? E.g., the "res" variable in that nootbook?

**Reply by rickyloynd-microsoft** [reply]
Track is kept by the simple rule above. But do you have a particular use in mind that wouldn't work?

**Reply by huangch** [reply]
Ok I guess now I get the point...It doesn't matter which role is assigned. The code just has to record which agent sent what message.

Thank you!

**Reply by nileshtrivedi** [reply]
Just ran into this, while following the tutorial of two comedians bouncing off each other and taking turns to respond: https://microsoft.github.io/autogen/docs/tutorial/introduction#roles-and-conversations

It's still not clear how to recover the agent names in result returned by `joe.initiate_chat`. Even though that's the exact output displayed in the tutorial. 😂 

<img width="833" alt="image" src="https://github.com/microsoft/autogen/assets/19304/c03f0279-c807-44b4-9eff-6b59c3b68b09">

**Reply by ekzhu** [reply]
@nileshtrivedi yes this is a known issue. Would you take a look at this PR addressing this issue? #2635

**Reply by nileshtrivedi** [reply]
@ekzhu I am not familiar with autogen internals. But yes, this would be desirable. The agents are instantiated with the names, so I had expected the chat_history to reflect those. For now, this is how I am massaging chat_history so that the conversation displayed in the UI actually makes sense:

```
    result = joe.initiate_chat(cathy, message="Cathy, tell me a joke.", max_turns=1)
    return [{"text": m["content"], "role": ("joe" if m["role"] == "assistant" else "cathy")} for m in result.chat_history]
```

**Reply by ekzhu** [reply]
Got it. Thanks for the tip
